pgjdbc/src/main/java/org/postgresql/util/internal/PgBufferedOutputStream.java

```diff
@@ -0,0 +1,194 @@
+/*
+ * Copyright (c) 2024, PostgreSQL Global Development Group
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.postgresql.util.internal;
+
+import org.postgresql.util.ByteConverter;
+import org.postgresql.util.GT;
+
+import java.io.EOFException;
+import java.io.FilterOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.Arrays;
+
+/**
+ * Buffered output stream. The key difference from {@link java.io.BufferedOutputStream} is that
+ * {@code PgBufferedOutputStream} does not perform synchronization.
+ * This is an internal class, and it is not meant to be used as a public API.
+ */
+public class PgBufferedOutputStream extends FilterOutputStream {
+  /**
+   * Buffer for the data
+   */
+  protected final byte[] buf;
+
+  /**
+   * Number of bytes stored in the buffer
+   */
+  protected int count;
+
+  public PgBufferedOutputStream(OutputStream out, int bufferSize) {
+    super(out);
+    buf = new byte[bufferSize];
+  }
+
+  protected void flushBuffer() throws IOException {
+    if (count > 0) {
+      out.write(buf, 0, count);
+      count = 0;
+    }
+  }
+
+  @Override
+  public void flush() throws IOException {
+    flushBuffer();
+    super.flush();
+  }
+
+  public void writeInt2(int val) throws IOException {
+    byte[] buf = this.buf;
+    if (buf.length - count < 2) {
+      flushBuffer();
+    }
+    int count = this.count;
+    ByteConverter.int2(buf, count, val);
+    this.count = count + 2;
+  }
+
+  public void writeInt4(int val) throws IOException {
+    byte[] buf = this.buf;
+    if (buf.length - count < 4) {
+      flushBuffer();
+    }
+    int count = this.count;
+    ByteConverter.int4(buf, count, val);
+    this.count = count + 4;
+  }
+
+  @Override
+  public void write(int b) throws IOException {
+    if (count >= buf.length) {
+      flushBuffer();
+    }
+    buf[count++] = (byte) b;
+  }
+
+  @Override
+  public void write(byte[] b, int off, int len) throws IOException {
+    if (len == 0) {
+      return;
+    }
+
+    if (count > 0) {
+      int avail = buf.length - count;
+      if (avail + buf.length <= len) {
```

Comment:
Is this branch really worthwhile?

Reply:
Sure. If the given buffer is very long, we do not want to copy from the given `b`. We would rather want to pass the given `b` as is to avoid extra copies. The stock `BufferedOutputStream` includes the similar optimization.